File: golang/pkg/agentbay/beta_network.go
package agentbay

import (
	"fmt"
	"strings"
	"time"

	"github.com/alibabacloud-go/tea/tea"
	mcp "github.com/aliyun/wuying-agentbay-sdk/golang/api/client"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/models"
)

// BetaNetworkResult wraps beta network bind token result and RequestID.
type BetaNetworkResult struct {
	models.ApiResponse
	Success      bool
	NetworkId    string
	NetworkToken string
	ErrorMessage string
}

// BetaNetworkStatusResult wraps beta network status result and RequestID.
type BetaNetworkStatusResult struct {
	models.ApiResponse
	Success      bool
	Online       bool
	ErrorMessage string
}

// BetaNetworkService provides beta methods to manage networks.
type BetaNetworkService struct {
	AgentBay *AgentBay
}

// BetaGetNetworkBindToken creates a network (or reuses provided networkId) and returns networkId + networkToken.
func (ns *BetaNetworkService) BetaGetNetworkBindToken(networkId string) (*BetaNetworkResult, error) {
	request := &mcp.CreateNetworkRequest{
		Authorization: tea.String("Bearer " + ns.AgentBay.APIKey),
	}
	if networkId != "" {
		request.NetworkId = tea.String(networkId)
	}

	maxAttempts := 3
	delay := 200 * time.Millisecond
	var resp *mcp.CreateNetworkResponse
	var err error
	for attempt := 1; attempt <= maxAttempts; attempt++ {
		resp, err = ns.AgentBay.Client.CreateNetwork(request)
		if err == nil {
			break
		}
		if attempt < maxAttempts {
			errStr := err.Error()
			if strings.Contains(errStr, "ServiceUnavailable") || strings.Contains(errStr, "503") {
				time.Sleep(delay)
				delay *= 2
				continue
			}
		}
	}

	requestID := models.ExtractRequestID(resp)
	if err != nil {
		return &BetaNetworkResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			ErrorMessage: fmt.Sprintf("Failed to create network: %v", err),
		}, nil
	}

	if resp == nil || resp.Body == nil {
		return &BetaNetworkResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			ErrorMessage: "Invalid response from CreateNetwork API",
		}, nil
	}

	body := resp.Body
	if body.Success == nil || !*body.Success {
		errorMsg := "Unknown error"
		if body.Message != nil {
			errorMsg = *body.Message
		}
		if body.Code != nil {
			errorMsg = fmt.Sprintf("[%s] %s", *body.Code, errorMsg)
		}
		return &BetaNetworkResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			ErrorMessage: errorMsg,
		}, nil
	}

	if body.Data == nil {
		return &BetaNetworkResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			ErrorMessage: "Network data not found in response",
		}, nil
	}

	createdNetworkID := ""
	if body.Data.NetworkId != nil {
		createdNetworkID = *body.Data.NetworkId
	}
	networkToken := ""
	if body.Data.NetworkToken != nil {
		networkToken = *body.Data.NetworkToken
	}

	return &BetaNetworkResult{
		ApiResponse:  models.ApiResponse{RequestID: requestID},
		Success:      true,
		NetworkId:    createdNetworkID,
		NetworkToken: networkToken,
		ErrorMessage: "",
	}, nil
}

// BetaDescribe queries beta network status (online/offline).
func (ns *BetaNetworkService) BetaDescribe(networkId string) (*BetaNetworkStatusResult, error) {
	if networkId == "" {
		return &BetaNetworkStatusResult{
			ApiResponse:  models.ApiResponse{RequestID: ""},
			Success:      false,
			Online:       false,
			ErrorMessage: "network_id is required",
		}, nil
	}

	request := &mcp.DescribeNetworkRequest{
		Authorization: tea.String("Bearer " + ns.AgentBay.APIKey),
		NetworkId:     tea.String(networkId),
	}

	maxAttempts := 3
	delay := 200 * time.Millisecond
	var resp *mcp.DescribeNetworkResponse
	var err error
	for attempt := 1; attempt <= maxAttempts; attempt++ {
		resp, err = ns.AgentBay.Client.DescribeNetwork(request)
		if err == nil {
			break
		}
		if attempt < maxAttempts {
			errStr := err.Error()
			if strings.Contains(errStr, "ServiceUnavailable") || strings.Contains(errStr, "503") {
				time.Sleep(delay)
				delay *= 2
				continue
			}
		}
	}

	requestID := models.ExtractRequestID(resp)
	if err != nil {
		errorStr := err.Error()
		if strings.Contains(errorStr, "NotFound") {
			return &BetaNetworkStatusResult{
				ApiResponse:  models.ApiResponse{RequestID: requestID},
				Success:      false,
				Online:       false,
				ErrorMessage: fmt.Sprintf("Network %s not found", networkId),
			}, nil
		}
		return &BetaNetworkStatusResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			Online:       false,
			ErrorMessage: fmt.Sprintf("Failed to describe network: %v", err),
		}, nil
	}

	if resp == nil || resp.Body == nil {
		return &BetaNetworkStatusResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			Online:       false,
			ErrorMessage: "Invalid response from DescribeNetwork API",
		}, nil
	}

	body := resp.Body
	if body.Success == nil || !*body.Success {
		errorMsg := "Unknown error"
		if body.Message != nil {
			errorMsg = *body.Message
		}
		if body.Code != nil {
			errorMsg = fmt.Sprintf("[%s] %s", *body.Code, errorMsg)
		}
		return &BetaNetworkStatusResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			Online:       false,
			ErrorMessage: errorMsg,
		}, nil
	}

	online := false
	if body.Data != nil && body.Data.Online != nil {
		online = *body.Data.Online
	}
	return &BetaNetworkStatusResult{
		ApiResponse:  models.ApiResponse{RequestID: requestID},
		Success:      true,
		Online:       online,
		ErrorMessage: "",
	}, nil
}

File: typescript/src/beta-network.ts
import { AgentBay } from "./agent-bay";
import * as $_client from "./api";
import { APIError } from "./exceptions";
import { extractRequestId } from "./types/api-response";
import type { NetworkResult, NetworkStatusResult } from "./types/api-response";
import { logAPICall, logAPIResponseWithDetails, logDebug } from "./utils/logger";

/**
 * Beta network service (trial feature).
 */
export class BetaNetworkService {
  private agentBay: AgentBay;

  constructor(agentBay: AgentBay) {
    this.agentBay = agentBay;
  }

  async getNetworkBindToken(networkId?: string): Promise<NetworkResult> {
    try {
      const request = new $_client.CreateNetworkRequest({
        authorization: `Bearer ${this.agentBay.getAPIKey()}`,
        loginRegionId: this.agentBay.getRegionId(),
        networkId,
      });

      logAPICall("CreateNetwork(beta)");
      if (networkId) {
        logDebug(`Request: NetworkId=${networkId}`);
      }

      const response = await this.agentBay.getClient().createNetwork(request);
      const requestId = extractRequestId(response) || "";

      if (!response.body || typeof response.body !== "object") {
        logAPIResponseWithDetails("CreateNetwork(beta)", requestId, false, undefined, "Invalid response format");
        return {
          requestId,
          success: false,
          networkId: "",
          networkToken: "",
          errorMessage: "Invalid response from CreateNetwork API",
        };
      }

      if (response.body.success === false) {
        const code = response.body.code ? `[${response.body.code}] ` : "";
        const msg = response.body.message || "Unknown error";
        return {
          requestId,
          success: false,
          networkId: "",
          networkToken: "",
          errorMessage: `${code}${msg}`,
        };
      }

      const data = response.body.data;
      if (!data || typeof data !== "object") {
        return {
          requestId,
          success: false,
          networkId: "",
          networkToken: "",
          errorMessage: "Network data not found in response",
        };
      }

      return {
        requestId,
        success: true,
        networkId: data.networkId || "",
        networkToken: data.networkToken || "",
      };
    } catch (error: unknown) {
      const msg = error instanceof Error ? error.message : `${error}`;
      throw new APIError(`CreateNetwork(beta) failed: ${msg}`);
    }
  }

  /**
   * Deprecated: use getNetworkBindToken().
   */
  async create(networkId?: string): Promise<NetworkResult> {
    return await this.getNetworkBindToken(networkId);
  }

  async describe(networkId: string): Promise<NetworkStatusResult> {
    if (!networkId) {
      return {
        requestId: "",
        success: false,
        online: false,
        errorMessage: "network_id is required",
      };
    }

    const maxAttempts = 3;
    let delayMs = 200;
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        const request = new $_client.DescribeNetworkRequest({
          authorization: `Bearer ${this.agentBay.getAPIKey()}`,
          networkId,
        });

        logAPICall("DescribeNetwork(beta)");
        logDebug(`Request: NetworkId=${networkId}`);

        const response = await this.agentBay.getClient().describeNetwork(request);
        const requestId = extractRequestId(response) || "";

        if (!response.body || typeof response.body !== "object") {
          logAPIResponseWithDetails("DescribeNetwork(beta)", requestId, false, undefined, "Invalid response format");
          return {
            requestId,
            success: false,
            online: false,
            errorMessage: "Invalid response from DescribeNetwork API",
          };
        }

        if (response.body.success === false) {
          const code = response.body.code ? `[${response.body.code}] ` : "";
          const msg = response.body.message || "Unknown error";
          return {
            requestId,
            success: false,
            online: false,
            errorMessage: `${code}${msg}`,
          };
        }

        const online = Boolean(response.body.data?.online);
        return {
          requestId,
          success: true,
          online,
        };
      } catch (error: unknown) {
        const msg = error instanceof Error ? error.message : `${error}`;
        if (attempt < maxAttempts && (msg.includes("ServiceUnavailable") || msg.includes("503"))) {
          await new Promise((resolve) => setTimeout(resolve, delayMs));
          delayMs *= 2;
          continue;
        }
        throw new APIError(`DescribeNetwork(beta) failed: ${msg}`);
      }
    }
    throw new APIError("DescribeNetwork(beta) failed: exhausted retries");
  }
}


